### DISCUSSION TITLE
renovate.yaml

### DISCUSSION THREAD

**Discussion by amezin** [renovatebot/renovate#29478] [Suggest an Idea]
### Tell us more.

Would it be possible to support yaml as the format of configuration and presets?

Writing complex regexes and jsonata transformations, and having to (double-)escape everything is incredibly painful (currently I have a few Python scripts to generate my Renovate configs and presets, because I can't escape everything correctly by hand). JSON5 doesn't seem to improve anything here.

Plus I'm not aware of any language server implementation for json5 (with schema support).

**Comment by HonkingGoose** [comment]
We had an old issue about this, with this comment from the lead maintainer:

> This topic came up again recently, so I will add a recent opinion: I don't want to support this unless a regular contributor to Renovate wants to add it and support it in the longer term. While one-off PRs are usually appreciated, in this case it would be like gifting a puppy - we'd be left behind to clean up after it and look after it all its life.

- https://github.com/renovatebot/renovate/issues/7031#issuecomment-1561757931

@rarkins is the comment above still your _current_ opinion on supporting the YAML format? As in: you're open for the _community_ to create (and then regularly _maintain_!) the YAML related code?

**Reply by rarkins** [reply]
I'm open to the idea of allowing yaml for config now. We could also allow configuration of the preferred fileName too

**Reply by HonkingGoose** [reply]
@rarkins Here's my draft for a new issue. I'll let you copy/paste/edit it into a proper clean actionable issue (or _issues_). :wink:

## Intro

The idea here is that users can opt-in to use the YAML format with Renovate in a:

- `renovate.yml` file, or
- `renovate.yaml` file

JSON remains the source of truth for the docs and code.

## Code todos

- [ ] Allow users to configure their preferred `fileName`.
- [ ] Add post-processor to docs build script that converts valid JSON examples to valid YAML.
- [ ] Any other things I'm forgetting to list here

### Docs: try content tabs with a single JSON tab first

- [ ] Keep using JSON in all our examples and code.
- [ ] Confirm we properly test for valid JSON examples in the docs.
- [ ] Enable and configure [Material for MkDocs's content tabs](https://squidfunk.github.io/mkdocs-material/reference/content-tabs/) feature, to prepare for later steps.
- [ ] Update docs to use Content Tabs for the current JSON, test docs keep working. Content tabs have a different format than our old plain JSON examples.
- [ ] Confirm linter correctly lints JSON in Content Tab layout.

### Add/write JSON-to-YAML converter, insert YAML into content tab in docs

- [ ] Add code that converts valid JSON configuration examples to YAML, insert the YAML in a content tab.
- [ ] Write tests to confirm conversion from JSON to YAML works as intended.
- [ ] Confirm linter correctly lints JSON in content tab.
- [ ] Confirm linter correctly lints YAML in content tab.
  - [ ] Use [Material for MkDocs's content tabs](https://squidfunk.github.io/mkdocs-material/reference/content-tabs/) feature to let users switch between JSON or YAML examples in the published docs.

### Related issues and discussions

- #29478
- #15193
- #10682
- #7031

**Reply by rarkins** [reply]
I've created https://github.com/renovatebot/renovate/issues/29537 which has some blockers. I had forgotten about https://github.com/renovatebot/renovate/issues/15193 when I created it, but maybe we can keep both for now

**Reply by HonkingGoose** [reply]
That's one long string of blocker issues!

You did _not_ create a issue about _Show YAML config examples in the docs_, is that intended? It might be good to capture my ideas about the docs in a proper issue, as this discussion will be hard to find later. :smile: Let me know if you want me to create the issue for you!

**Reply by rarkins** [reply]
I think it should be part of the last issue

**Reply by HonkingGoose** [reply]
This is the last issue in the string of blockers:

- https://github.com/renovatebot/renovate/issues/27972

Should the _Show YAML config examples in the docs_ be a part of that issue then?

Edit: I copy/pasted todos into that issue.
